src/f32ext/fract.rs

```diff
@@ -0,0 +1,89 @@
+use super::copysign;
+use super::utils;
+use super::utils::FloatComponents;
+/// Floating point whole number for a single-precision float.
+use core::f32;
+use core::u32;
+
+pub(super) fn fract_sign(x: f32) -> f32 {
+    let x_bits: u32 = x.to_bits();
+    let exponent: i32 = x.extract_exponent_value();
+    // we know it is *only* fraction
+    if exponent < 0_i32 {
+        return x;
+    }
+    let exponent_clamped = i32::max(exponent, 0_i32) as u32;
+
+    // find the part of the fraction that would be left over
+    let fractional_part: u32 = x_bits.overflowing_shl(exponent_clamped).0 & utils::MANTISSA_MASK;
+    // if there isn't a fraction we can just return 0
+    if fractional_part == 0_u32 {
+        // most people don't actually care about -0.0, so would it be better to just not copysign?
+        return copysign::copysign(0.0_f32, x);
+    }
+    //alternatively use -1.0? subtraction would surely be more costly though right?
+    let exponent_shift: u32 = (fractional_part.leading_zeros() - (32_u32 - 23_u32)) + 1;
+    let fractional_normalized: u32 =
+        fractional_part.overflowing_shl(exponent_shift).0 & utils::MANTISSA_MASK;
+
+    let new_exponent_bits = (utils::EXPONENT_BIAS - (exponent_shift))
+        .overflowing_shl(23_u32)
+        .0;
+    copysign::copysign(f32::from_bits(fractional_normalized | new_exponent_bits), x)
+}
+
+// don't want to delete maybe this is the better way?
```

Comment:
What is this specifically?

Reply:
The bit manipulation I did at the end of fract can be replaced with a subtraction of `1.0_f32`, I'm not sure whether a subtraction would be worse or better than the bit manipulation I made, particularly in terms of performance where this library would be used.  It appears as if the result would be the same either way. It simplifies the code as well.